src/compiler/core.ts

```diff
@@ -350,80 +367,271 @@ namespace ts {
+     *
+     * @param source A map-like from which properties should be copied.
+     * @param target A map-like to which properties should be copied.
+     */
+    export function copyOwnProperties<T>(source: MapLike<T>, target: MapLike<T>): void {
+        for (const key in source) if (hasOwnProperty.call(source, key)) {
+            target[key] = source[key];
         }
-        return result;
     }
 
-    export function forEachValue<T, U>(map: MapLike<T>, callback: (value: T) => U): U {
-        let result: U;
-        for (const id in map) {
-            if (result = callback(map[id])) break;
+    /**
+     * Reduce the properties of a map.
+     *
+     * NOTE: This is intended for use with Map<T> objects. For MapLike<T> objects, use
+     *       reduceOwnProperties instead as it offers better runtime safety.
+     *
+     * @param map The map to reduce
+     * @param callback An aggregation function that is called for each entry in the map
+     * @param initial The initial value for the reduction.
+     */
+    export function reduceProperties<T, U>(map: Map<T>, callback: (aggregate: U, value: T, key: string) => U, initial: U): U {
+        let result = initial;
+        for (const key in map) {
+            result = callback(result, map[key], String(key));
         }
         return result;
     }
 
-    export function forEachKey<T, U>(map: MapLike<T>, callback: (key: string) => U): U {
-        let result: U;
-        for (const id in map) {
-            if (result = callback(id)) break;
+    /**
+     * Reduce the properties defined on a map-like (but not from its prototype chain).
+     *
+     * NOTE: This is intended for use with MapLike<T> objects. For Map<T> objects, use
+     *       reduceProperties instead as it offers better performance.
+     *
+     * @param map The map-like to reduce
+     * @param callback An aggregation function that is called for each entry in the map
+     * @param initial The initial value for the reduction.
+     */
+    export function reduceOwnProperties<T, U>(map: MapLike<T>, callback: (aggregate: U, value: T, key: string) => U, initial: U): U {
+        let result = initial;
+        for (const key in map) if (hasOwnProperty.call(map, key)) {
+            result = callback(result, map[key], String(key));
         }
         return result;
     }
 
-    export function lookUp<T>(map: MapLike<T>, key: string): T {
-        return hasProperty(map, key) ? map[key] : undefined;
+    /**
+     * Counts the properties of a map.
+     *
+     * NOTE: This is intended for use with Map<T> objects. For MapLike<T> objects, use
+     *       countOwnProperties instead as it offers better runtime safety.
+     *
+     * @param map A map whose properties should be counted.
+     * @param predicate An optional callback used to limit which properties should be counted.
+     */
+    export function countProperties<T>(map: Map<T>, predicate?: (value: T, key: string) => boolean) {

```

Comment:
Are all of these new helper functions actually used somewhere? Just making sure we don't carry around dead weight.